keps/sig-network/20190603-EndpointSlice-API.md

```diff
@@ -0,0 +1,380 @@
+- Provide better load balancing for K8s service backends.
+
+## Proposal
+
+### EndpointSlice API
+The following new EndpointSlice API will be added to the `Discovery` API group.
+
+```
+type EndpointSlice struct {
+    metav1.TypeMeta `json:",inline"`
+    // OwnerReferences should be set when the object is derived from a k8s
+    // object.
+    // The object labels may contain the following key:
+    // kubernetes.io/service: the label value indicates the name of the
+    // service from which the EndpointSlice is derived from.
+    // +optional
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+    // Metadata of the Endpoints in the EndpointSlice.
+    EndpointMeta `json:",inline"`
+    // Endpoints of this EndpointSlice.
+    // +optional
+    // +listType=atomic
+    Endpoints []Endpoint `json:"endpoints,omitempty" protobuf:"bytes,2,opt,name=endpoints"`
+}
+
+type Endpoint struct {
+    // Required: Targets of the endpoint. Must contain at least one target.
+    // Different consumers (e.g. kube-proxy) handle different types of
+    // targets in the context of its own capabilities.
+    // +listType=set
+    Targets []string `json:"targets,omitempty" protobuf:"bytes,1,opt,name=targets"`
+    // The Hostname of this endpoint.
+    // +optional 
+    Hostname *string `json:"hostname,omitempty" protobuf:"bytes,2,opt,name=hostname"`
+    // Required: the conditions of the endpoint.
+    Conditions EndpointConditions  `json:"conditions,omitempty" protobuf:"bytes,3,opt,name=conditions"`
+    // Reference to object providing the endpoint.
+    // +optional 
+    TargetRef *v1.ObjectReference `json:"targetRef,omitempty" protobuf:"bytes,4,opt,name=targetRef"`
+    // Topology can contain arbitrary topology information associated with the
+    // endpoint.
+    // Key/value pairs contained in topology must conform with the label format.
+    // https://kubernetes.io/docs/concepts/overview/working-with-objects/labels
+    // Topology may contain the following well known keys:
+    // kubernetes.io/hostname: the value indicates the hostname of the node
+    // where the endpoint is located. This should match the corresponding
+    // node label.
+    // topology.kubernetes.io/zone: the value indicates the zone where the
+    // endpoint is located. This should match the corresponding node label.
+    // topology.kubernetes.io/region: the value indicates the region where the
+    // endpoint is located. This should match the corresponding node label.
+    // +optional 
+    Topology map[string]string `json:"topology,omitempty" protobuf:"bytes,5,opt,name=topology"`
+}
+
+type EndpointConditions struct {
+    // Ready indicates if the endpoint is ready to serve traffic.
+    Ready bool `json:"ready,omitempty" protobuf:"bytes,1,opt,name=ready"`
+}
+
+type EndpointMeta struct {
+    // This field specifies the list of ports associated with each endpoint in the EndpointSlice
```

Comment:
...If this list is empty or unspecified, no ports or protocols are are defined and the endpoints are effectively unreachable.

Do we want to allow that, or demand that at least one ports block, even if `{}` be provided?